plugins/preserve-tombs.cpp

```diff
@@ -0,0 +1,269 @@
+        set_config_val(config, CONFIG_CYCLES, 25);
+    }
+
+    is_enabled = get_config_bool(config, CONFIG_IS_ENABLED);
+    cycle_freq = get_config_val(config, CONFIG_CYCLES);
+    DEBUG(config,out).print("loading persisted enabled state: %s\n",
+                            is_enabled ? "true" : "false");
+
+    if (is_enabled) update_tomb_assignments(out);
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (event == DFHack::SC_WORLD_UNLOADED) {
+        tomb_assignments.clear();
+        if (is_enabled) {
+            DEBUG(config,out).print("world unloaded; disabling %s\n",
+                                    plugin_name);
+            is_enabled = false;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    if (is_enabled && world->frame_counter - cycle_timestamp >= cycle_freq)
+        update_tomb_assignments(out);
+    return CR_OK;
+}
+// </BOILERPLATE>
+
+
+
+
+// On unit death - check if we assigned them a tomb
+//
+//
+void onUnitDeath(color_ostream& out, void* ptr) {
+    // input is void* that contains the unit id
+    int32_t unit_id = reinterpret_cast<std::intptr_t>(ptr);
+
+    // check if unit was assigned a tomb in life
+    auto it = tomb_assignments.find(unit_id);
+    if (it == tomb_assignments.end()) return;
+
+    // assign that unit to their previously assigned tomb in life
+    int32_t building_id = it->second;
+    if (!assign_to_tomb(unit_id, building_id)) return;
+
+    // success, print status update and remove assignment from our memo-list
+    out.print("Unit %d died - assigning them to tomb %d\n", unit_id, building_id);
+    tomb_assignments.erase(it);
+
+}
+
+
+// Update tomb assignments
+//
+//
+static void update_tomb_assignments(color_ostream &out) {
+
+    // check tomb civzones for assigned units
+    for (auto* bld : world->buildings.other.ZONE_TOMB) {
+
+        auto* tomb = virtual_cast<df::building_civzonest>(bld);
+        if (!tomb || !tomb->flags.bits.exists) continue;
+        if (!tomb->assigned_unit) continue;
+        if (Units::isDead(tomb->assigned_unit)) continue; // we only care about living units
+
+        auto it = tomb_assignments.find(tomb->assigned_unit_id);
+
+        if (it == tomb_assignments.end()) {
+            tomb_assignments.emplace(tomb->assigned_unit_id, tomb->id);
+            out.print("%s new tomb assignment, unit %d to tomb %d\n", plugin_name, tomb->assigned_unit_id, tomb->id);
+        }
+
+        else {
+            if (it->second != tomb->id) {
+                out.print("%s tomb assignment to %d changed, (old: %d, new: %d)\n", plugin_name, tomb->assigned_unit_id, it->second, tomb->id);
+            }
+            it->second = tomb->id;
+        }
+
+    }
+
+    // now check our civzones for unassignment / deleted zone
+    std::erase_if(tomb_assignments,[&](const auto& p){
```

Comment:
don't you want to do this block *before* the `check tomb civzones for assigned units` block? what if I've swapped assignments between two tombs?

Reply:
The previous block that iterates over the `TOMB_ZONE` vector would already have assigned them correctly so they wouldn't be erased here.